3scale/porta
# Add Stripe payment form

**What this PR does / why we need it**:

For any failed invoice payment (using Stripe) in the developer portal, we need to add a checkout form to enable developers retry payment. 

- [x] Improve getting the client secret for the invoice (@guicassolato)
- [x] Review UX (@cntlsn)
- [x] Add success page (redirecting to show page)
- [x] Update Docs to explain how to update show liquid template (@porueesq) (doing [here](https://issues.redhat.com/browse/THREESCALE-6657))

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-6623

**Verification steps** 

To test this PR you will need a [Stripe](https://stripe.com) test merchant account with the corresponding _Secret Key_ and  _Publishable Key_. Check the [docs](https://stripe.com/docs/keys) if needed.

You will also need a Stripe _Webhook Signing Secret_. In case your environment is reachable through the Internet and with a valid TLS certificate, you can obtain the signing secret by creating a Webhook in the Stripe Dashboard (details in the steps below). Otherwise, you need to download and install the [Stripe CLI](https://stripe.com/docs/stripe-cli).
 
**1. In Admin portal:**
- Go to **Account Settings > Personal > Tokens**
- Create a **Read & Write** token with the **Billing API** scope
- If your are running 3scale in an environment publicly reachable and with a verifiable TLS certificate:
    - Go the **Stripe Dashboard**
    - Go to **Developer > Webhooks**
    - Click **Add endpoint**
    - Fill in with the following endpoint URL: `https://<provider-admin-domain>/api/payment_callbacks/stripe_callbacks?access_token=<value-of-access-token>`
    - In **Events to send**, add "payment_intent.succeeded"
    - Click **Save**
    - Now **Click to reveal** the **Signing secret** of the webhook just created and take note of it
- If your are NOT running 3scale in an environment publicly reachable or the TLS certificate cannot be verified:
    - Launch the Stripe CLI listening to `http(s)://<provider-admin-domain>/api/payment_callbacks/stripe_callbacks?access_token=<value-of-access-token>`
    - Take note of the "webhook signing secret" that the Stripe CLI prints in the terminal
- Now back in the **3scale Admin Portal**, Go to **Audience > Billing > Charging & Gateway**
- Select **Charging enabled** and click **Save**
- In **Credit card gateway > Gateway**, select **Stripe**
- Add your Secret Key, Publishable Key and Webhook Signing Secret
- Click "Save"

**2. In Developer Portal:**
- Login with a developer account
- Go to **Settings > Credit Card Details**
- Add a test credit card
- Use `4000002760003184`, as this card requires authentication on all transactions, for `MM/YY` use any future date and any `CVC`. See [docs](https://stripe.com/docs/testing#regulatory-cards)

**3. In Admin portal:**
- Go to **Audience > Accounts > Listing**
- Select the same account whose credit card info was added in the Dev Portal before
- At the top, follow the breadcrumb link to **invoices**
- Click **Create invoice**
- Click on the generated invoice to edit it
- Click the **+ Add button**, to add a line item
- Put any _Name_, _Quantity_ and _Description_
- Click on **Create Line Item**
- Click on _Issue invoice_
- Click on _Charge_

Payment will fail. That is expected, as the selected test credit card required authentication.

**4. In Admin Portal:**

- If you are working with a **newly created instance** of the Developer Portal, skip this step.
- If you want to update an **existing instance of the Developer Portal**, do the following:
  - Go to **Audience > Developer Portal > Content**
  - Edit **Root > Invoices > Show** template
  - Make this change:
<img width="1005" alt="Screenshot 2021-02-04 at 16 36 53" src="https://user-images.githubusercontent.com/13486237/106916431-72008200-6707-11eb-940e-9ffc387ac449.png">


**5. In Developer Portal:**
- Go to **Settings > Invoices**
- The invoice you just created should appear as **pending**
- Click on **Show**
- Click on **Pay Invoice**
- Add a new card number, date and CVC (you can use "4111 1111 1111 1111")
- Click **Pay**

**Screenshots:**

> <img width="870" alt="Screenshot 2021-02-04 at 16 39 31" src="https://user-images.githubusercontent.com/13486237/106916572-9fe5c680-6707-11eb-8998-84fa49b7678c.png">


> ![invoice02](https://user-images.githubusercontent.com/13486237/106611359-29f72900-6568-11eb-8c87-965665c6280a.png)
